
# <font color=red>**TED Talks**</font>

[Kaggle Competition](https://www.kaggle.com/rounakbanik/ted-talks)

[GitHub - Carlos Scovino](https://github.com/cscovino/TED-Talks-Analysis)


###  <font color=blue> **Archivo**:</font>  ted_main.csv


| Tipo | Columna | Descripcion |
| :--- | :--- | :--- |
| Numeric | comments | The number of first level comments made on the talk |
| Text | description | A blurb of what the talk is about |
| Numeric | duration | The duration of the talk in seconds |
| Text | event | The TED/TEDx event where the talk took place |
| Timestamp | film_date | The Unix timestamp of the filming |
| Numeric | languages | The number of languages in which the talk is available |
| Text | main_speaker | The first named speaker of the talk |
| Text | name | The official name of the TED Talk. Includes the title and the speaker. |
| Numeric | num_speaker | The number of speakers in the talk |
| Timestamp | published_date | The Unix timestamp for the publication of the talk on TED.com |
| Text | ratings | A stringified dictionary of the various ratings given to the talk (inspiring, fascinating, jaw dropping, etc.) |
| Text | related_talks | A list of dictionaries of recommended talks to watch next |
| Text | speaker_occupation | The occupation of the main speaker |
| Text | tags | The themes associated with the talk |
| Text | title | The title of the talk |
| Text | url | The URL of the talk |
| Numeric | views | The number of views on the talk |

###  <font color=blue> **Archivo**:</font>  transcripts.csv

| Tipo | Columna | Descripcion |
| :--- | :--- | :--- |
| Text | transcript | The official English transcript of the talk. |
| Text | url | The URL of the talk |

<div class="alert alert-block alert-info">

# Imports

</div>

In [139]:
import pandas as pd

import ast
import datetime

<div class="alert alert-block alert-info">

# Dataframes

</div>

In [140]:
df_main = pd.read_csv('./dataset/ted_main.csv',sep=",",quotechar='"',quoting=1)
df_transcript = pd.read_csv('./dataset/transcripts.csv',sep=",",quotechar='"',quoting=1)

<div class="alert alert-block alert-info">

# Describes

</div>

### Data: Main

In [141]:
df_main.columns

Index(['comments', 'description', 'duration', 'event', 'film_date',
       'languages', 'main_speaker', 'name', 'num_speaker', 'published_date',
       'ratings', 'related_talks', 'speaker_occupation', 'tags', 'title',
       'url', 'views'],
      dtype='object')

In [142]:
df_main.count()[0]

2550

In [143]:
df_main.describe()

comments     duration     film_date    languages  num_speaker  \
count  2550.000000  2550.000000  2.550000e+03  2550.000000  2550.000000   
mean    191.562353   826.510196  1.321928e+09    27.326275     1.028235   
std     282.315223   374.009138  1.197391e+08     9.563452     0.207705   
min       2.000000   135.000000  7.464960e+07     0.000000     1.000000   
25%      63.000000   577.000000  1.257466e+09    23.000000     1.000000   
50%     118.000000   848.000000  1.333238e+09    28.000000     1.000000   
75%     221.750000  1046.750000  1.412964e+09    33.000000     1.000000   
max    6404.000000  5256.000000  1.503792e+09    72.000000     5.000000   

       published_date         views  
count    2.550000e+03  2.550000e+03  
mean     1.343525e+09  1.698297e+06  
std      9.464009e+07  2.498479e+06  
min      1.151367e+09  5.044300e+04  
25%      1.268463e+09  7.557928e+05  
50%      1.340935e+09  1.124524e+06  
75%      1.423432e+09  1.700760e+06  
max      1.506092e+09  4.722711e+07

In [144]:
df_main.head(2)

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   

     event   film_date  languages  main_speaker  \
0  TED2006  1140825600         60  Ken Robinson   
1  TED2006  1140825600         43       Al Gore   

                                        name  num_speaker  published_date  \
0  Ken Robinson: Do schools kill creativity?            1      1151367060   
1       Al Gore: Averting the climate crisis            1      1151367060   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   

                                       related_talks speaker_occupation  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   Climate advocate   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   

                         title  \
0  Do schools kill creativity?   
1  Averting the climate crisis   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520

<div class="alert alert-block alert-success">

## Ratings

</div>

In [145]:
lst_ratings = ["Funny", "Beautiful", "Ingenious",
               "Courageous", "Longwinded", "Confusing",
               "Informative", "Fascinating", "Unconvincing",
               "Persuasive", "Jaw-dropping", "OK",
               "Obnoxious", "Inspiring"]

In [146]:
df_main['ratings'][2]

"[{'id': 7, 'name': 'Funny', 'count': 964}, {'id': 3, 'name': 'Courageous', 'count': 45}, {'id': 9, 'name': 'Ingenious', 'count': 183}, {'id': 1, 'name': 'Beautiful', 'count': 60}, {'id': 21, 'name': 'Unconvincing', 'count': 104}, {'id': 11, 'name': 'Longwinded', 'count': 78}, {'id': 8, 'name': 'Informative', 'count': 395}, {'id': 10, 'name': 'Inspiring', 'count': 230}, {'id': 22, 'name': 'Fascinating', 'count': 166}, {'id': 2, 'name': 'Confusing', 'count': 27}, {'id': 25, 'name': 'OK', 'count': 146}, {'id': 24, 'name': 'Persuasive', 'count': 230}, {'id': 23, 'name': 'Jaw-dropping', 'count': 54}, {'id': 26, 'name': 'Obnoxious', 'count': 142}]"

###  <font color=blue> **DataFrame**:</font>  df_ratings

| ix | rating_funny | rating_beautiful | rating_ingenious | ... | rating_<14> | 
| :--- | :--- | :--- | :--- | --- | --- |
| 0 | 19645 | 4573 | 6073 | ... | ... |
| 1 | 544 | 58 | 56 | ... | ... |
| 2 | 964 | 60 | 183 | ... | ... | 
| ... | ...| ... | ... | ... | ... |
| 2549 | ...| ... | ... | ... | ... |

In [147]:
all_ratings = []

df_ratings = df_main['ratings']

for ix, ratings in df_ratings.items():
    
    rec_rating = {}
    rec_rating['ix'] = ix
    
    # String to Python Data Type
    ratings_lst = ast.literal_eval(ratings)

    for rating in ratings_lst:   
        rec_rating["rating_"+rating['name'].lower()]=rating['count']

    all_ratings.append(rec_rating)

In [148]:
df_ratings =  pd.DataFrame(all_ratings,
                           index=[ dic['ix'] for dic in all_ratings],
                           columns=["rating_funny","rating_beautiful","rating_ingenious",
                                    "rating_courageous","rating_longwinded","rating_confusing",
                                    "rating_informative","rating_fascinating","rating_unconvincing",
                                    "rating_persuasive","rating_jaw-dropping","rating_ok",
                                    "rating_obnoxious","rating_inspiring"] )

In [149]:
len(df_ratings)

2550

In [150]:
df_ratings.shape

(2550, 14)

In [151]:
df_ratings.head()

rating_funny  rating_beautiful  rating_ingenious  rating_courageous  \
0         19645              4573              6073               3253   
1           544                58                56                139   
2           964                60               183                 45   
3            59               291               105                760   
4          1390               942              3202                318   

   rating_longwinded  rating_confusing  rating_informative  \
0                387               242                7346   
1                113                62                 443   
2                 78                27                 395   
3                 53                32                 380   
4                110                72                5433   

   rating_fascinating  rating_unconvincing  rating_persuasive  \
0               10581                  300              10704   
1                 132                  258                268   
2                 166                  104                230   
3                 132                   36                460   
4                4606                   67               2542   

   rating_jaw-dropping  rating_ok  rating_obnoxious  rating_inspiring  
0                 4439       1174               209             24924  
1                  116        203               131               413  
2                   54        146               142               230  
3                  230         85                35              1070  
4                 3736        248                61              2893

<div class="alert alert-block alert-success">

## Tags

</div>

###  <font color=blue> **DataFrame**:</font>  df_tags

| ix | tag_children | tag_creativity | tag_culture | ... | tag_other |
| :--- | :--- | :--- | :--- | --- | --- |
| 0 | 1 | 1 | 1 | ... | 1 |
| 1 | 0 | 0 | 0 | ... | 3 |
| 2 | 0 | 0 | 0 | ... | 2 |
| ... | ...| ... | ... | ... | ... |
| 2549 | ...| ... | ... | ... | ... |

In [152]:
df_main['tags'].head()

0    ['children', 'creativity', 'culture', 'dance',...
1    ['alternative energy', 'cars', 'climate change...
2    ['computers', 'entertainment', 'interface desi...
3    ['MacArthur grant', 'activism', 'business', 'c...
4    ['Africa', 'Asia', 'Google', 'demo', 'economic...
Name: tags, dtype: object

In [185]:
lst_tags = ["technology", "science", "design", "business", "collaboration", "innovation", "social_change",
            "health", "nature", "environment", "future", "communication", "activism", "children",
            "personal_growth", "humanity", "society", "identity", "community", "culture", "global_issues",
            "entertainment", "art", "politics", "economics", "religion"]

In [186]:
all_tags = []
df_tags = df_main['tags']

for ix, tags in df_tags.items():

    # --- Empty Dict ---
    rec_tag = {}
    rec_tag['ix'] = ix
    for tag in lst_tags:
        rec_tag["tag_"+tag.lower().replace(" ","_")]=0
    rec_tag['tag_other']=0
    
    # --- Update Dict ---
    tags = ast.literal_eval(tags)
    for tag in tags:
        
        if "tag_"+tag.lower() in rec_tag.keys():
            rec_tag["tag_"+tag.lower().replace(" ","_")]+=1
        else:
            rec_tag["tag_other"]+=1
            
    all_tags.append(rec_tag)

In [190]:
columns=[ "tag_"+tag.lower() for tag in lst_tags]
columns.append("tag_other")
print(columns)

['tag_technology', 'tag_science', 'tag_design', 'tag_business', 'tag_collaboration', 'tag_innovation', 'tag_social_change', 'tag_health', 'tag_nature', 'tag_environment', 'tag_future', 'tag_communication', 'tag_activism', 'tag_children', 'tag_personal_growth', 'tag_humanity', 'tag_society', 'tag_identity', 'tag_community', 'tag_culture', 'tag_global_issues', 'tag_entertainment', 'tag_art', 'tag_politics', 'tag_economics', 'tag_religion', 'tag_other']


In [193]:
df_tags =  pd.DataFrame(all_tags,
                        index=[ row['ix'] for row in all_tags ],
                        columns=columns)

In [194]:
df_tags.head()

tag_technology  tag_science  tag_design  tag_business  tag_collaboration  \
0               0            0           0             0                  0   
1               1            1           0             0                  0   
2               1            0           0             0                  0   
3               0            0           0             1                  0   
4               0            0           0             0                  0   

   tag_innovation  tag_social_change  tag_health  tag_nature  tag_environment  \
0               0                  0           0           0                0   
1               0                  0           0           0                1   
2               0                  0           0           0                0   
3               0                  0           0           0                1   
4               0                  0           1           0                0   

   ...  tag_identity  tag_community  tag_culture  tag_global_issues  \
0  ...             0              0            1                  0   
1  ...             0              0            1                  0   
2  ...             0              0            0                  0   
3  ...             0              0            0                  0   
4  ...             0              0            0                  0   

   tag_entertainment  tag_art  tag_politics  tag_economics  tag_religion  \
0                  0        0             0              0             0   
1                  0        0             0              0             0   
2                  1        0             0              0             0   
3                  0        0             1              0             0   
4                  0        0             0              1             0   

   tag_other  
0          5  
1          5  
2          7  
3          5  
4          9  

[5 rows x 27 columns]

In [195]:
len(df_tags)

2550

In [196]:
df_tags.shape

(2550, 27)

<div class="alert alert-block alert-success">

## Film & Published Dates

</div>

In [24]:
f = lambda x: datetime.datetime.fromtimestamp(int(x)).strftime('%d-%m-%Y')

df_main['film_date'] = df_main['film_date'].apply(f)
df_main['published_date'] = df_main['published_date'].apply(f)

In [25]:
df_main.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  24-02-2006         60   Ken Robinson   
1  TED2006  24-02-2006         43        Al Gore   
2  TED2006  23-02-2006         26    David Pogue   
3  TED2006  25-02-2006         35  Majora Carter   
4  TED2006  21-02-2006         48   Hans Rosling   

                                            name  num_speaker published_date  \
0      Ken Robinson: Do schools kill creativity?            1     26-06-2006   
1           Al Gore: Averting the climate crisis            1     26-06-2006   
2                  David Pogue: Simplicity sells            1     26-06-2006   
3             Majora Carter: Greening the ghetto            1     26-06-2006   
4  Hans Rosling: The best stats you've ever seen            1     27-06-2006   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

### Data: Transcripts

In [19]:
df_transcript.count()[0]

2467

In [6]:
df_transcript.columns

Index(['transcript', 'url'], dtype='object')

In [17]:
df_transcript.head()

transcript  \
0  Good morning. How are you?(Laughter)It's been ...   
1  Thank you so much, Chris. And it's truly a gre...   
2  (Music: "The Sound of Silence," Simon & Garfun...   
3  If you're here today — and I'm very happy that...   
4  About 10 years ago, I took on the task to teac...   

                                                 url  
0  https://www.ted.com/talks/ken_robinson_says_sc...  
1  https://www.ted.com/talks/al_gore_on_averting_...  
2  https://www.ted.com/talks/david_pogue_says_sim...  
3  https://www.ted.com/talks/majora_carter_s_tale...  
4  https://www.ted.com/talks/hans_rosling_shows_t...